File: algorithms/trivial_structs/linked_list.py
# Linked List

# Time Complexity:
# append():     O(1)
# prepend():    O(1)
# delete():     O(n)
# find():       O(n)
# get_size():   O(1)
#
# Space Complexity:
# Memory:       O(n)

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    # O(1)
    def append(self, data):
        node = Node(data)

        if not self.head:
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node

        self.size += 1

    # O(1)
    def prepend(self, data):
        node = Node(data)

        if not self.head:
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head = node

        self.size += 1

    # O(n)
    def delete(self, data):
        if not self.head:
            return

        if self.head.data == data:
            self.head = self.head.next
            self.size -= 1

            if not self.head:
                self.tail = None
            return

        prev = self.head
        curr = self.head.next

        while curr:
            if curr.data == data:
                prev.next = curr.next

                if curr == self.tail:
                    self.tail = prev

                self.size -= 1
                return

            prev, curr = curr, curr.next

    # O(n)
    def find(self, data):
        curr = self.head

        while curr:
            if curr.data == data:
                return curr
            curr = curr.next

        return None

    def get_size(self):
        return self.size

File: cpp_algorithms/trivial_structs/linked_list.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

// Linked List
//
// Time Complexity:
// append():     O(1)
// prepend():    O(1)
// delete():     O(n)
// find():       O(n)
// get_size():   O(1)
//
// Space Complexity:
// Memory:       O(n)

class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        next = nullptr;
    }
};

class LinkedList {
private:
    Node* head;
    Node* tail;
    int size;

public:
    LinkedList() {
        head = nullptr;
        tail = nullptr;
        size = 0;
    }

    // O(1)
    void append(int data) {
        Node* node = new Node(data);

        if (head == nullptr) {
            head = tail = node;
        } else {
            tail->next = node;
            tail = node;
        }

        size++;
    }

    // O(1)
    void prepend(int data) {
        Node* node = new Node(data);

        if (head == nullptr) {
            head = tail = node;
        } else {
            node->next = head;
            head = node;
        }

        size++;
    }

    // O(n)
    void delete_node(int data) {
        if (head == nullptr)
            return;

        if (head->data == data) {
            Node* temp = head;

            head = head->next;
            delete temp;

            size--;

            if (head == nullptr)
                tail = nullptr;

            return;
        }

        Node* prev = head;
        Node* curr = head->next;

        while (curr != nullptr) {
            if (curr->data == data) {
                prev->next = curr->next;

                if (curr == tail)
                    tail = prev;

                delete curr;
                size--;

                return;
            }

            prev = curr;
            curr = curr->next;
        }
    }

    // O(n)
    Node* find(int data) {
        Node* curr = head;

        while (curr != nullptr) {
            if (curr->data == data)
                return curr;

            curr = curr->next;
        }

        return nullptr;
    }

    // O(1)
    int get_size() {
        return size;
    }

    ~LinkedList() {
        Node* curr = head;

        while (curr != nullptr) {
            Node* next = curr->next;
            delete curr;
            curr = next;
        }
    }
};
